File: Toggly.FeatureManagement.Flutter/toggly/test/entity_gate_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:feature_flags_toggly/src/entity_gate.dart';

void main() {
  final datetimeGate = EntityGate(
    requirement: 'all',
    rules: [
      EntityGateRule(
        property: 'BirthDate',
        op: 'gt',
        value: '2026-01-01',
        type: 'datetime',
      ),
    ],
  );

  tearDown(clearRegisteredContexts);

  test('detects entity gates and fails closed', () {
    expect(isEntityGate(true), isFalse);
    expect(isEntityGate(datetimeGate), isTrue);
    expect(resolveEvaluatedDefinition(datetimeGate), isFalse);
    expect(resolveEvaluatedDefinition(null, defaultValue: true), isTrue);
    expect(resolveEvaluatedDefinition(false, defaultValue: true), isFalse);
    expect(resolveEvaluatedDefinition(datetimeGate, defaultValue: true), isFalse);
  });

  test('evaluates datetime gt and flattens snapshots', () {
    expect(
      resolveEvaluatedDefinition(
        datetimeGate,
        context: TogglyEntityContext(
          kind: 'Puppy',
          key: '1',
          attributes: {'BirthDate': '2026-06-15T00:00:00Z'},
        ),
      ),
      isTrue,
    );
    expect(
      toBooleanDefinitions({'On': true, 'Off': false, 'Gated': datetimeGate}),
      {'On': true, 'Off': false, 'Gated': false},
    );
  });

  test('operators and fail-closed branches', () {
    final anyGate = EntityGate(
      requirement: 'any',
      rules: [
        EntityGateRule(property: 'Color', op: 'eq', value: 'red'),
        EntityGateRule(property: 'Color', op: 'eq', value: 'blue'),
      ],
    );
    expect(applyEntityGate(anyGate, {'Color': 'blue'}), isTrue);
    expect(
      applyEntityGate(EntityGate(requirement: 'all', rules: anyGate.rules), {'Color': 'blue'}),
      isFalse,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Color', op: 'neq', value: 'red'),
        ]),
        {},
      ),
      isFalse,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Code', op: 'gt', value: '9'),
        ]),
        {'Code': '10'},
      ),
      isFalse,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'color', op: 'eq', value: 'RED'),
        ]),
        {'Color': 'red'},
      ),
      isTrue,
    );
    expect(applyEntityGate(EntityGate(requirement: 'all', rules: []), {'Color': 'red'}), isFalse);
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Age', op: 'gte', value: '2', type: 'number'),
        ]),
        {'Age': 2},
      ),
      isTrue,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Color', op: 'in', value: 'red, blue'),
        ]),
        {'Color': 'BLUE'},
      ),
      isTrue,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Name', op: 'contains', value: 'pup'),
        ]),
        {'Name': 'Puppy'},
      ),
      isTrue,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Tags', op: 'contains', value: 'beta', type: 'string[]'),
        ]),
        {'Tags': ['GA', 'Beta']},
      ),
      isTrue,
    );
    expect(
      applyEntityGate(
        EntityGate(requirement: 'all', rules: [
          EntityGateRule(property: 'Color', op: 'matches', value: 'red'),
        ]),
        {'Color': 'red'},
      ),
      isFalse,
    );
  });

  test('parses mixed JSON definitions', () {
    final parsed = parseEvaluatedDefinitions({
      'On': true,
      'Gated': {
        'requirement': 'all',
        'rules': [
          {'property': 'Color', 'op': 'eq', 'value': 'red'},
        ],
      },
    });
    expect(toBooleanDefinitions(parsed)['Gated'], isFalse);
    expect(
      resolveEvaluatedDefinition(
        parsed['Gated'],
        context: TogglyEntityContext(kind: 'Puppy', key: '1', attributes: {'Color': 'red'}),
      ),
      isTrue,
    );
  });

  test('registerContext is local only', () {
    registerContext('Puppy', (entity) {
      final puppy = entity as Map;
      return TogglyEntityContext(
        kind: 'Puppy',
        key: puppy['id'] as String,
        attributes: {'Color': puppy['color']},
      );
    });
    expect(mapEntityContext('Puppy', {'id': '1', 'color': 'red'})?.key, '1');
    expect(resolveEntityContext('Kitten', {'id': '1'}), isNull);
  });
}

File: Toggly.FeatureManagement.Android/toggly-android-core/src/test/kotlin/io/toggly/core/EntityGateTest.kt
package io.toggly.core

import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class EntityGateTest {
    private val datetimeGate = EntityGate(
        requirement = "all",
        rules = listOf(
            EntityGateRule(property = "BirthDate", op = "gt", value = "2026-01-01", type = "datetime")
        )
    )

    @After
    fun tearDown() {
        clearRegisteredContexts()
    }

    @Test
    fun `detects entity gates`() {
        assertFalse(isEntityGate(true))
        assertTrue(isEntityGate(datetimeGate))
    }

    @Test
    fun `fails closed without context`() {
        assertFalse(resolveEvaluatedDefinition(EvaluatedDefinition.Gate(datetimeGate)))
    }

    @Test
    fun `falls back to default for absent definition`() {
        assertFalse(resolveEvaluatedDefinition(null))
        assertTrue(resolveEvaluatedDefinition(null, defaultValue = true))
        assertFalse(resolveEvaluatedDefinition(EvaluatedDefinition.BooleanValue(false), defaultValue = true))
        assertFalse(resolveEvaluatedDefinition(EvaluatedDefinition.Gate(datetimeGate), defaultValue = true))
    }

    @Test
    fun `evaluates datetime gt locally`() {
        val enabled = resolveEvaluatedDefinition(
            EvaluatedDefinition.Gate(datetimeGate),
            TogglyEntityContext("Order", "1", mapOf("BirthDate" to "2026-06-15T00:00:00Z"))
        )
        assertTrue(enabled)
    }

    @Test
    fun `flattens mixed definitions`() {
        val flattened = toBooleanDefinitions(
            mapOf(
                "On" to EvaluatedDefinition.BooleanValue(true),
                "Off" to EvaluatedDefinition.BooleanValue(false),
                "Gated" to EvaluatedDefinition.Gate(datetimeGate)
            )
        )
        assertEquals(mapOf("On" to true, "Off" to false, "Gated" to false), flattened)
    }

    @Test
    fun `evaluates any all requirements`() {
        val gate = EntityGate(
            requirement = "any",
            rules = listOf(
                EntityGateRule("Color", "eq", "red"),
                EntityGateRule("Color", "eq", "blue")
            )
        )
        assertTrue(applyEntityGate(gate, mapOf("Color" to "blue")))
        assertFalse(applyEntityGate(gate.copy(requirement = "all"), mapOf("Color" to "blue")))
    }

    @Test
    fun `fails closed on missing attr unknown op empty rules`() {
        assertFalse(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Color", "neq", "red"))),
                emptyMap()
            )
        )
        assertFalse(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Color", "matches", "red"))),
                mapOf("Color" to "red")
            )
        )
        assertFalse(applyEntityGate(EntityGate("all", emptyList()), mapOf("Color" to "red")))
        assertFalse(applyEntityGate(EntityGate("any", emptyList()), mapOf("Color" to "red")))
    }

    @Test
    fun `does not treat string ordered compares as numbers`() {
        assertFalse(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Code", "gt", "9"))),
                mapOf("Color" to "10", "Code" to "10")
            )
        )
    }

    @Test
    fun `compares equality in contains numbers datetimes case-insensitively`() {
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("color", "eq", "RED"))),
                mapOf("Color" to "red")
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Age", "gte", "2", "number"))),
                mapOf("Age" to 2)
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Age", "lt", "2", "number"))),
                mapOf("Age" to "1.5")
            )
        )
        assertFalse(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Age", "lte", "2", "number"))),
                mapOf("Age" to 3)
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Born", "gt", "2026-01-01", "datetime"))),
                mapOf("Born" to "2026-06-01T00:00:00Z")
            )
        )
        assertFalse(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Born", "gt", "not-a-date", "datetime"))),
                mapOf("Born" to "2026-06-01T00:00:00Z")
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Color", "in", "red, blue"))),
                mapOf("Color" to "BLUE")
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Name", "contains", "ord"))),
                mapOf("Name" to "Order")
            )
        )
        assertTrue(
            applyEntityGate(
                EntityGate("all", listOf(EntityGateRule("Tags", "contains", "beta", "string[]"))),
                mapOf("Tags" to listOf("GA", "Beta"))
            )
        )
    }

    @Test
    fun `parses mixed definitions json`() {
        val raw = """{"On":true,"Gated":{"requirement":"all","rules":[{"property":"Color","op":"eq","value":"red"}]}}"""
        val parsed = parseEvaluatedDefinitions(raw)
        assertTrue(parsed["On"] is EvaluatedDefinition.BooleanValue)
        assertTrue(parsed["Gated"] is EvaluatedDefinition.Gate)
        assertEquals(mapOf("On" to true, "Gated" to false), toBooleanDefinitions(parsed))
        assertEquals(
            true,
            resolveEvaluatedDefinition(
                parsed["Gated"],
                TogglyEntityContext("Order", "1", mapOf("Color" to "red"))
            )
        )
    }

    @Test
    fun `registerContext maps entities locally`() {
        registerContext("Order") { order: Map<String, String> ->
            TogglyEntityContext("Order", order.getValue("id"), mapOf("Color" to order["color"]))
        }
        val mapped = mapEntityContext("Order", mapOf("id" to "1", "color" to "red"))
        assertEquals("1", mapped?.key)
        assertNull(resolveEntityContext("Kitten", mapOf("id" to "1")))
    }
}
